src/mpid/ch3/channels/nemesis/src/ch3_progress.c

```diff
@@ -38,6 +38,15 @@ static MPIDI_CH3_PktHandler_Fcn *pktArray[PKTARRAY_SIZE];
 #define MPIDI_POSTED_RECV_DEQUEUE_HOOK(x) 0
 #endif
 
+#ifdef MPICH_IS_THREADED
+/* Because we use MPIDI_CH3I_progress_blocked to make other threads that need progress
```

Comment:
What problem is this fixing? Timeouts?

Reply:
Yes.

reference: https://github.com/pmodels/mpich/pull/4385#issuecomment-597112772

Reply:
Oh, sorry. This is for nemesis. No, it isn't fixing any testing failures. It is for consistency.

Reply:
(Ignore my previous mumbling :smile: ) In ch4, due to significant overhead in progress, often runs beyond a thousand cycles, a contending thread at `pthread_mutex_lock` is often put to sleep. I guess a sleeping thread is only waking up at certain kernel clock interrupt. When `MPL_thread_yield` is a noop -- as current default -- the contending thread will nearly never able to grab the lock from a progress_wait loop. So a fix there (ch4) is to ensure a more pronounced `MPL_thread_yield`. But as we do that, `ch3/nemesis` start to suffer. Due to nemesis's design, a progress test is super quick and there isn't a lock monopoly issue as in ch4. And if we introduce a more heavy `MPL_thread_yield`, the performance of ch3/nemesis suffers greatly. So in ch3/nemesis, we really want the `MPL_thread_yield` to be a noop. 

This commit ensures the `MPL_thread_yield` is a noop in nemesis, so we are free to experiment and use a heavier and effective yield in ch4.

Reply:
Back to your question. Currently, ch3/nemesis doesn't suffer from timeouts. This commit is to ensure it stays this way when we introduce changes to `MPL_thread_yield` later.

Reply:
OK, my question then is why limit this hack to just progress?

Reply:
> Back to your question. Currently, ch3/nemesis doesn't suffer from timeouts. This commit is to ensure it stays this way when we introduce changes to `MPL_thread_yield` later.

Then lets drop it from this PR and add it back if it's actually needed.

Reply:
> OK, my question then is why limit this hack to just progress?

Yield is only used in progress.

> Then lets drop it from this PR and add it back if it's actually needed.

Note that the commit itself has nothing to do with ch4. It is a refactor to isolate the special need for ch3/nemesis (that require thread_yield to be a noop) from ch4. 

While adding it back to ch4 changes doesn't make sense, I would agree that we can split this commit into a separate PR. Redefine `MPL_thread_yield` is hacky. The proper fix is to define its own `THREAD_CS_YIELD`. But that entangles with the use of recursive mutex and certainly involves more code changes. -- Maybe it makes sense to discuss it in a separate PR.